three-cards-divination/src/main/java/com/cyber/mysticism/tarot/service/ThreeCardsDivinationService.java

```diff
@@ -87,4 +131,38 @@ private List<Card> extractCardsWithFilter(String duplicatedCardName) {
         Stream<Card> cards = getCards().stream();
         return cards.filter(card -> !duplicatedCardName.equals(card.name())).collect(toShuffledStream()).limit(2).toList();
     }
+
+    /**
+     * @return a map containing a unique reading for a particular user in a recursive manner
+     */
+    private Map<String, Card> getUniqueReadingBasedOnHashCode(TarotUser retrievedUser) throws DivinationException {
+        Map<String, Card> threeCardsDivinationReading = getReading();
+        String hashCode = getHashCode(threeCardsDivinationReading);
+
+        if (threeCardsDivinationRepository.countReadingsByHashCodeForUsername(hashCode, retrievedUser.getUsername()) != 0) {
+            // If duplicated, recursively call the method to get a new reading
+            logger.info("event=reading_duplicated_for_user, type={}, user={}", THREE_CARDS_DIVINATION, retrievedUser.getUsername());
+            return getUniqueReadingBasedOnHashCode(retrievedUser);
+        }
+
+        saveReadingForUser(retrievedUser, hashCode, threeCardsDivinationReading);
+
+        return threeCardsDivinationReading;
+    }
+
+    /**
+     * Save new hash_code and corresponding reading to DB
+     */
+    private void saveReadingForUser(TarotUser retrievedUser, String hashCode, Map<String, Card> threeCardsDivinationReading) {
+        Set<ThreeCardsDivinationReading> retrievedUserReadings = retrievedUser.getReadings();
+        retrievedUserReadings.add(new ThreeCardsDivinationReading(hashCode, convertReadingMapToJsonString(threeCardsDivinationReading)));
+        retrievedUser.setReadings(retrievedUserReadings);
+        userRepository.save(retrievedUser);
+
+        logger.info("event=reading_for_user_success, type={}, user={}", THREE_CARDS_DIVINATION, retrievedUser.getUsername());
+    }
+
+    private String getHashCode(Map<String, Card> reading) {
+        return String.valueOf(Objects.hashCode(reading));
+    }
```

Comment:
**Suggestion:** Consider using `Objects.hash(Object...)` instead of `Objects.hashCode(Object)`. The `Objects.hash(Object...)` method can accept multiple parameters and will generate a hash code based on the values of all the parameters. [best practice]
```suggestion
    private String getHashCode(Map<String, Card> reading) {
        return String.valueOf(Objects.hash(reading));
    }
```